test_unishox2.c

```diff
@@ -117,6 +118,27 @@ int test_ushx_cd(char *input, int preset) {
   perc *= 100;
   printf("%s: %d/%d=", input, clen, len);
   printf("%.2f%%\n", perc);
+
+  // test terminator
+  for (int i = 0; i < ((byte*)&terminator)[3];) {
+    cbuf[clen++] = ((byte*)&terminator)[i++];
+  }
+  cbuf[clen++] = cbuf[0];
+  cbuf[clen++] = cbuf[1];
+  cbuf[clen++] = cbuf[2];
+
+  memset(dbuf, 0, sizeof(dbuf));
+  dlen = unishox2_decompress_preset_lines(cbuf, sizeof cbuf, dbuf, preset, NULL);
+  dbuf[dlen] = '\0';
+  if (dlen != len) {
+    printf("Fail len (term): %d, %d:\n%s\n%s\n", len, dlen, input, dbuf);
+    return 0;
+  }
+  if (strncmp(input, dbuf, len)) {
+    printf("Fail cmp (term):\n%s\n%s\n", input, dbuf);
+    return 0;
+  }
+
```

Comment:
I appreciate the idea of testing the terminator this way.  I really like it, but is there a way for doing it without changing the API?   According to my understanding, the new parameter only serves the purposes of writing unit test cases?

Reply:
Not only for ut. The developer can choose a different package frame for the compressed bytes, e.g., appending the term codes to the output, and transport the byte sequence `compressed data || term codes` insteadof `length || compressed data`.

Reply:
I also want to keeping the api, but if one prefer `length || compressed data` pattern, the term codes are wasted, so it should be easy for developers to distinguish which bytes are term codes. If we always append the term codes, it seems hard to calculate the term code position.

Reply:
I understand what you are saying.  Can we have a boolean parameter for terminator instead of `long *` to say whether they want the full terminator in the output or not.  In that case we don't need to split the PR.  However, I think the code for doing this and adding the Terminator for preset 1 can be simpler.

Reply:
it is ok, but there are still some corner cases, for example the developer want select a short one between prefix length and suffix term codes, using a boolean switch, he has to compress input twice.

Reply:
This library is good only for short strings.  If we use `varint` for the prefix length, 1 byte can be used for a length of 128 and 2 bytes for a length of 16384, so length can always be prefixed without having to worry much about which is longer.

Usually the length is stored somewhere.  So I think needing full terminator is rare, so having separate buffer for terminator code is more work for developer and makes difficult to understand the API.

Reply:
The length of term codes, after this patch, varies from 0 to 3 bytes depending on the H-codes preset and `state`, so is has some room for preferring the suffix term codes pattern. But it it rare really. So I will try another way to passing out the term codes with out changing the api. The proposal would be like this: when the compressor api returns `len` and `len < output buffer size`, then the `cbuf[0..len-1]` are the normal compressed data, and `cbuf[len]` contains a special term indicator. To recover the full term codes, developers are required providing a big enough output buffer, and expand the left term code bytes from the indicator byte with a new api. Pro is that the previous api signatures are not changed, in the meanwhile, cons is that the behavior of compressor api changed a little, writing one more byte beyond the return bytes, then if the output buffer size is not limit, the program would write over the buffer boundary.